### PR TITLE
chore(framework): Initial commit

### PR DESCRIPTION
**Pull request by adrien2p**
**What**
- Initiate the framework package (which is just a place to move things around for now)
- move the config loader and related resources as well as the `ConfigModule` type
- Create a ConfigManager singleton which prepare and store the config (later can be stored entirely in the container) and allow for easier test override
- re export the logger from the framework
- replace medusa config loader with the framework one
- `build` run type check on tests as well but `prepublishOnly` will not fail on build if tests are typed broken

FIXES FRMW-2607
FIXES FRMW-2609
FIXES FRMW-2614
FIXES FRMW-2618

### CODE REVIEW

**Comment by adrien2p** [comment]
This pr include experimental test of `nodenext` resolution to use sub path exports/imports. If the test fail and create more issues then we will have to post pone the usage of sub paths and rely on classic approach

File: integration-tests/modules/package.json [hunk 1]
```diff
@@ -5,8 +5,8 @@
   "license": "MIT",
   "private": true,
   "scripts": {
-    "test:integration": "jest --silent=false --no-cache --maxWorkers=50% --bail --detectOpenHandles --forceExit --logHeapUsage",
-    "test:integration:chunk": "jest --silent --no-cache --bail --maxWorkers=50% --forceExit --testPathPattern=$(echo $CHUNKS | jq -r \".[${CHUNK}] | .[]\")",
+    "test:integration": "NODE_OPTIONS=--experimental-vm-modules jest --silent=false --no-cache --maxWorkers=50% --bail --detectOpenHandles --forceExit --logHeapUsage",
```
Comment (thetutlage) [inline integration-tests/modules/package.json:8]: Just curious about the role of `experimental-vm-modules` here

**Comment by thetutlage** [comment]
So far it looks good to me!

File: integration-tests/modules/package.json (hunk 1 above)
Comment (adrien2p) [reply integration-tests/modules/package.json:8]: Jest complaining about esm modules https://jestjs.io/docs/ecmascript-modules

File: packages/framework/framework/src/config/config.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
+import { ConfigModule } from "./types"
+import { deepCopy, isDefined } from "@medusajs/utils"
+import { Logger } from "@medusajs/types"
+
+export class ConfigManager {
+  /**
+   * A flag to secify if we are in product or not, determine weither an error would be critical and thrown or just logged as a warning in developement
```
Comment (olivermrbl) [inline packages/framework/framework/src/config/config.ts]:
```suggestion
   * A flag to secify if we are in production or not, determine whether an error would be critical and thrown or just logged as a warning in developement
```

File: packages/framework/framework/src/config/config.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+  ): ConfigModule["projectConfig"] {
+    const outputConfig: ConfigModule["projectConfig"] = deepCopy(projectConfig)
+
+    if (!outputConfig?.redisUrl) {
+      console.log(`redisUrl not found. A fake redis instance will be used.`)
+    }
+
+    outputConfig.http = this.buildHttpConfig(projectConfig)
+
+    let workedMode = outputConfig?.workerMode!
```
Comment (olivermrbl) [inline packages/framework/framework/src/config/config.ts]: nit:
```suggestion
    let workerMode = outputConfig?.workerMode!
```

**Review by olivermrbl — APPROVED** [review]
LGTM, nice work!

**Pull request merged**